### DISCUSSION TITLE
Has anyone gotten Apple to notarize something with OSHI in it?

### DISCUSSION THREAD

**Discussion by BrianAtAustin** [oshi/oshi#2042] [Q&A] [answered]
If so, how?

For the last several months I've been working on a diagnostic package creation service for our product. If the user has an error, we direct them to run this service, which grabs all the logs, configs, etc, and zips them up so that they can email them to us. We want to include hardware information as well, so I added OSHI to the system and it worked great.

The problem is that -- when trying to notarize the MAC code, apple is complaining that the some of the included binaries aren't signed. (Included below are snippets of the failure.json from apple). The problem isn't with OSHI, but with the libjnidispatch.jnilibs  for MAC. 

Given that OSHI works (very well) on a MAC, I'm hoping that somebody knows a workaround.

Our other option is to try and strip out those libs and just disable this functionality on MAC (and keep it on windows), so a follow on question -- are those libraries only used by macs? What happens if I strip them out?

Thanks in Advance!

 {
      "severity": "error",
      "code": null,
      "path": "(our product)com/sun/jna/darwin-aarch64/libjnidispatch.jnilib",
      "message": "The binary is not signed with a valid Developer ID certificate.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "(our product)/com/sun/jna/darwin-aarch64/libjnidispatch.jnilib",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "arm64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "(our product)/com/sun/jna/darwin-x86-64/libjnidispatch.jnilib",
      "message": "The binary is not signed.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "(our product)/com/sun/jna/darwin-x86-64/libjnidispatch.jnilib",
      "message": "The signature does not include a secure timestamp.",
      "docUrl": null,
      "architecture": "x86_64"x
    },

**Comment by dbwiddis** [comment, marked as answer]
libjnidispatch.jnilib is the native JNI stub that is a key part of the JNA project which is a key dependency and integral part of how OSHI accesses native code. In short, no it wouldn't work.

In theory we could replace all the native code with command line parsing, but that's ugly.  Also in theory we can use the future foreign function code upcoming in the JDK to drop JNA as a dependency.  That's planned but it's also at least a year away (JDK21).

Asking about signing this jnilib would be appropriate for the [jna-users mailing list](https://groups.google.com/g/jna-users).

**Comment by BrianAtAustin** [comment]
Thank you for the answer. I will look at the jna-users lists.

**Comment by dbwiddis** [comment]
An option: Build the native bits yourself, sign them, and replace the jnilib:  https://github.com/java-native-access/jna/tree/master/native

**Reply by BrianAtAustin** [reply]
After some investigation and discussion, we ended up doing exactly that. Thank you for your suggestions and help, and I do appreciate the work on OSHI and the responsiveness of this forum!
